src/api/spec/controllers/webui/package_controller_spec.rb

```diff
@@ -397,4 +397,82 @@
       end
     end
   end
+
+  describe "GET #revisions" do
+    before do
+      User.current = user
+    end
+
+    context "without source access" do
+      before do
+        package.add_flag("sourceaccess", "disable")
+        package.save
+        get :revisions, project: source_project, package: package
+      end
+
+      it { expect(flash[:error]).to eq("Could not access revisions") }
+      it { expect(response).to redirect_to(package_show_path(project: source_project.name, package: package.name)) }
+    end
+
+    context "with source access" do
+      before do
+        get :revisions, project: source_project, package: package
+      end
+
+      after do
+        # This is necessary to delete the commits created in the before statement
+        package.destroy
+      end
+
+      it { expect(assigns(:project)).to eq(source_project) }
+      it { expect(assigns(:package)).to eq(package) }
+
+      context "with no revisions" do
+        it { expect(assigns(:lastrev)).to eq(1) }
+        it { expect(assigns(:revisions)).to eq([1]) }
+      end
+
+      context "with less than 21 revisions" do
+        let(:package_with_commits) { create(:package_with_file, name: "package_with_commits", project: source_project) }
+
+        before do
+          19.times { |i| Suse::Backend.put("/source/#{source_project}/#{package_with_commits}/somefile.txt", i.to_s) }
+          get :revisions, project: source_project, package: package_with_commits
+        end
+
+        after do
+          # This is necessary to delete the commits created in the before statement
+          package_with_commits.destroy
+        end
+
+        it { expect(assigns(:lastrev)).to eq(20) }
+        it { expect(assigns(:revisions)).to eq((1..20).to_a.reverse) }
+      end
+
+      context "with 21 revisions" do
+        let(:package_with_more_commits) { create(:package_with_file, name: "package_with_more_commits", project: source_project) }
+
+        before do
+          20.times { |i| Suse::Backend.put("/source/#{source_project}/#{package_with_more_commits}/somefile.txt", i.to_s) }
```

Comment:
That's just a thought, but we could move this piece to a new package_with_revisions factory and give it a trait for number of revisions. I think that would allow us to remove the after blocks.